Adding unit testing for various cases of cluster transition

anvddriesch opened the pull request:
Towards: https://github.com/giantswarm/giantswarm/issues/15468

This adds a unit test for `computeCreateClusterStatusConditions`. We can specify the condition and version history of the cluster to see how the conditions are set in different scenarios.

## Checklist

- [ ] Update changelog in CHANGELOG.md.

anvddriesch commented on service/controller/resource/statuscondition/create_test.go:
```diff
@@ -20,36 +19,239 @@ import (
+			},
+			operatorVersion: "8.7.7",
+
+			// TODO: this should be "updating" but it does not work yet
```
As we saw in our clusters recently, the updating condition can not appear twice.

xh3b4sd commented:
It is pretty cool that you found a creative solution for the tests. Not saying you should fix all of this, I just want to provide some points from the top of my head about the ideal scenario. 

- Unit tests are most awesome if they do not require mocking. Simpler is better. Here we need to use the fake clients and create objects and whatnot. This makes it more complicated. 
- Table driven tests are most useful if there is a clear input/output definition, at best in complete form. With the tests here we only seem to check the last values of a list but do not assert the full list during state changes. The current representation makes it also harder to comprehend the state change behaviour.

anvddriesch commented:
Hey, thank you for the feedback.
The test in its current form already existed including the mocking and table structure. My work was adding functionality by making it possible to change various values of the input to easily add scenarios. The other possibility would have been to rewrite everything from scratch but I thought it would take a lot of time. The function we test here is very long with a lot of inputs. I would prefer to have it split into separate functions which then can have simple unit tests.

xh3b4sd commented:
Agree the function is complex. Having the tests in place as is helps already. As a next step I would have tried to change the interface to improve the situation for the future.

anvddriesch commented:
I've been thinking how to make this better and had this approach in mind:
- Split the decision logic of `computeCreateClusterStatusConditions`  into separate functions (with just primitive inputs and outputs) while using the current version of the test to ensure the functionality does not change
- Add simple tests for the individual functions.
-  Fix the problems in `computeCreateClusterStatusConditions`

That's quite a refactoring effort for a component we might not keep for that long, though. So I'm interested in your approach. You say "change the interface" but I'm not exactly sure what you are referring to. Which interface, and how would you change it?

xh3b4sd commented:
I think of it in terms of input and output without any magic in between. There should only be some algorithmic processing in memory without having to worry about network or third parties. I would define what needs to get in and see what I would want to come out of it. The input is everything related to the current and desired state. The output may simply be the CR status as a whole or parts of it. I would also move the CR status processing to its own package. Packages in go make it super simple to make a clear cut. After such a refactoring the unit tests should be simpler and the coverage bigger without any magic in between.

anvddriesch commented:
With the change in `apiextensions` we can now see the rollback correctly in the status history.
https://github.com/giantswarm/apiextensions/pull/687

As you can see, this cluster was upgraded and then downgraded again.
```
    conditions:
    - condition: Updated
      lastTransitionTime: "2021-02-18T10:10:30Z"
    - condition: Updating
      lastTransitionTime: "2021-02-18T09:42:44Z"
    - condition: Updated
      lastTransitionTime: "2021-02-18T09:30:03Z"
    - condition: Updating
      lastTransitionTime: "2021-02-18T09:09:56Z"
    - condition: Created
      lastTransitionTime: "2021-02-18T08:59:24Z"
    - condition: Creating
      lastTransitionTime: "2021-02-18T08:44:56Z"
    id: so4d8
    versions:
    - lastTransitionTime: "2021-02-18T10:10:30Z"
      version: 9.3.1
    - lastTransitionTime: "2021-02-18T09:30:03Z"
      version: 9.3.1-fix
    - lastTransitionTime: "2021-02-18T08:59:24Z"
      version: 9.3.1
  provider:
    network:
      cidr: 10.1.7.0/24
      vpcID: vpc-0b0bd81e8e41816a4
```
As soon as `apiextensions` is released I'll update the dependency here and then I think we are good to go with this.
